Add support for ed25519ph user keys in hashedrekord

ret2libc opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary
<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->
Add support for Ed25519ph user keys in hashedrekord type, so we can test #1724 better with 2 options (ecdsa+sha256 and ed25519ph).

This change requires https://github.com/sigstore/sigstore/pull/1595 and https://github.com/sigstore/rekor/pull/1959 .

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->
* Added `x509.NewSignatureWithOpts` to support ED25519ph
* Allow ED25519 public keys when validating hashedrekord entries

#### Documentation
<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

Hayden-IO commented on pkg/types/hashedrekord/v0.0.1/hashedrekord_v0_0_1_schema.json:
```diff
@@ -38,7 +38,7 @@
                             "description": "The hashing function used to compute the hash value",
                             "type": "string",
-                            "enum": [ "sha256" ]
+                            "enum": [ "sha256", "sha512" ]
```
This will be the first time we're adding a non-sha256 hash, so we'll want to double check and test we don't hardcode it elsewhere.

Hayden-IO commented on pkg/types/hashedrekord/v0.0.1/entry.go:
```diff
@@ -160,11 +159,6 @@ func (v *V001Entry) validate() (pki.Signature, pki.PublicKey, error) {
 		return nil, nil, types.ValidationError(err)
 	}
 
-	_, isEd25519 := keyObj.CryptoPubKey().(ed25519.PublicKey)
```
Just for context, when I wrote up https://github.com/sigstore/rekor/issues/1325, my intention was to support ed25519ph only for hashedrekord, and ed25519 for all other types. This might mean making some changes to the 'pkg/pki/x509/x509' library to support loading in a custom verifier like on https://github.com/sigstore/rekor/blob/70d6663e297068383f708155f4017110c6f46713/pkg/pki/x509/x509.go#L87

ret2libc replied:
What do you think about [this](https://github.com/sigstore/rekor/pull/1945/commits/2d93686760f8fe40ea552ca9895c2922bfcb7780#diff-1cad0eed4767a1ef5dbac388d3e45fc4fe1a3c4fb6187f572040b515d81660b8) way of customizing the verifier?

ret2libc commented on pkg/types/hashedrekord/v0.0.1/entry_test.go:
```diff
@@ -93,7 +93,7 @@ func TestCrossFieldValidation(t *testing.T) {
 	dataSHA := hex.EncodeToString(h[:])
 
-	signer, _ := signature.LoadSigner(key, crypto.SHA256)
+	signer, _ := signature.LoadSigner(key, crypto.SHA256, signature.LoadDefaultSV, nil)
```
Probably we can avoid all these changes if we add a `LoadSignerWithOpts`/`LoadVerifierWithOpts`/`LoadSignerVerifierWithOpts` instead of changing the existing ones.

ret2libc committed:
Update sigstore/sigstore and use LoadOptions

Signed-off-by: Riccardo Schirone <riccardo.schirone@trailofbits.com>

ret2libc committed:
Added support for Ed25519ph in HashedRekord entries

- Made X509 Signatures configurable with LoadOptions
- Removed existing check that limited the use of Ed25519 keys in
  HashedRekord entries
- Used Ed25519ph Signer/Verifier for HashedRekord entries

Signed-off-by: Riccardo Schirone <riccardo.schirone@trailofbits.com>

ret2libc commented on pkg/types/hashedrekord/v0.0.1/entry.go:
```diff
@@ -230,10 +238,11 @@ func (v V001Entry) CreateFromArtifactProperties(_ context.Context, props types.A
 		return nil, errors.New("only one public key must be provided")
 	}
 
+	hashAlgorithm, hashValue := util.UnprefixSHA(props.ArtifactHash)
```
this is to allow passing as `props.ArtifactHash` either a regular sha value, or a prefixed sha value, which can be used to determine what is the hash algorithm used.

Is this a breaking change? cc @haydentherapper

bobcallaway commented on go.mod:
```diff
@@ -204,3 +204,5 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+replace github.com/sigstore/sigstore => github.com/trail-of-forks/sigstore v0.0.0-20240117103256-3095d93bafe8
```
this would need to be removed before merging.

woodruffw replied:
Yep! I believe this is here until https://github.com/sigstore/sigstore/pull/1595 gets merged (correct me if I'm wrong @ret2libc), so it probably makes sense to merge #1958 first and then we can rebase here 🙂

Hayden-IO reviewed (commented):
LGTM, this looks good.

While this might be obvious from the change since it only modifies hashedrekord, i wanted to double check that this won't affect other types. rekord also uses x509.NewSignature indirectly through a factory, but given that we've added a new signature function rather than modify the existing one, this has no impact on other types.

Let's merge https://github.com/sigstore/rekor/pull/1958 first (or split this into two) if that works for you, just to make it easier to see the two different changes.

Hayden-IO reviewed (commented):
LGTM, thanks for splitting this PR up. Will approve once we merge s/s and remove the replace.

Hayden-IO commented on go.mod:
```diff
@@ -204,3 +204,6 @@ require (
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+// TODO: REMOVE ME
```
good to remove now

Hayden-IO reviewed (commented):
LGTM, thanks for your patience and for all the discussion on the underlying verifier interface. This came out really clean!

Can you update the test for this (https://github.com/sigstore/rekor/blob/main/pkg/types/hashedrekord/v0.0.1/entry_test.go#L116-L120 and https://github.com/sigstore/rekor/blob/main/pkg/types/hashedrekord/v0.0.1/entry_test.go#L195-L211), which I would have expected to be failing now (so it might not have been testing the right thing)?

ret2libc committed:
Fix entry tests to ensure ed25519 is now accepted

Signed-off-by: Riccardo Schirone <riccardo.schirone@trailofbits.com>

Hayden-IO commented on go.mod:
```diff
@@ -26,7 +26,7 @@ require (
 	github.com/sassoftware/relic v7.2.1+incompatible
 	github.com/secure-systems-lab/go-securesystemslib v0.8.0
-	github.com/sigstore/sigstore v1.8.1
+	github.com/sigstore/sigstore v1.8.2-0.20240215213640-922069a3bc39
```
can you also bump to v1.8.2, we just cut a release

woodruffw replied:
Sure, doing now!
